src/Workspaces/Core/Portable/Utilities/BKTree.Builder.cs

```diff
@@ -0,0 +1,286 @@
+                _characterSpans = new TextSpan[distinctValues.Length];
+
+                var characterIndex = 0;
+                for (int i = 0; i < distinctValues.Length; i++)
+                {
+                    var value = distinctValues[i];
+                    _characterSpans[i] = new TextSpan(characterIndex, value.Length);
+
+                    foreach (var ch in value)
+                    {
+                        _allLowerCaseCharacters[characterIndex] = char.ToLower(ch);
+                        characterIndex++;
+                    }
+                }
+
+                // We will have one node for each string value that we are adding.
+                _builderNodes = new BuilderNode[distinctValues.Length];
+                _compactEdges = new Edge[distinctValues.Length * CompactEdgeAllocationSize];
+            }
+
+            internal BKTree Create()
+            {
+                for (var i = 0; i < _characterSpans.Length; i++)
+                {
+                    Add(_characterSpans[i], insertionIndex: i);
+                }
+
+                var nodes = ImmutableArray.CreateBuilder<Node>(_builderNodes.Length);
+
+                // There will be one less edge in the graph than nodes.  Each node (except for the
+                // root) will have a single edge pointing to it.
+                var edges = ImmutableArray.CreateBuilder<Edge>(Math.Max(0, _builderNodes.Length - 1));
+
+                BuildArrays(nodes, edges);
+
+                return new BKTree(_allLowerCaseCharacters, nodes.MoveToImmutable(), edges.MoveToImmutable());
+            }
+
+            private void BuildArrays(ImmutableArray<Node>.Builder nodes, ImmutableArray<Edge>.Builder edges)
+            {
+                var currentEdgeIndex = 0;
+                for (var i = 0; i < _builderNodes.Length; i++)
+                {
+                    var builderNode = _builderNodes[i];
+                    var edgeCount = builderNode.EdgeCount;
+
+                    nodes.Add(new Node(builderNode.CharacterSpan, edgeCount, currentEdgeIndex));
+
+                    if (edgeCount > 0)
+                    {
+                        if (edgeCount <= CompactEdgeAllocationSize)
+                        {
+                            // When tehre are less than 4 elements, copy from teh _compact array.
+                            var start = i * CompactEdgeAllocationSize;
+                            var end = start + edgeCount;
+                            for (var j = start; j < end; j++)
+                            {
+                                edges.Add(_compactEdges[j]);
+                            }
+                        }
+                        else
+                        {
+                            // When there are more than 4 elements, then we have to manually 
+                            // copy over the spilled items ourselves.
+                            var spilledEdges = builderNode.SpilloverEdges;
```

Comment:
Above you said "If it needs more than that space, then we have a fall-over dictionary that it can store information in," and the combination of that explanation and the name "SpilloverEdges" made me think you might still use the 4 in-array locations and then the *spillover* would go into the dictionary. If "spillover" is the technical term for this sort of thing, then I was just ignorant of that. If it's not, then this might all be a bit misleading.